Fixed issue 402. Now simc programs with unbalanced brackets will raise an error.

Chasmiccoder opened the pull request:
Fixed #402
now the program gives an error if the brackets are not balanced. Checks for simple parenthesis '(', square brackets '[', and flower brackets '{' as well.
@Math-O5 @frankhart2018

Math-O5 commented:
@Chasmiccoder Well done!

Math-O5 commented:
@Chasmiccoder Why did you create a function that will loop through all source_code again? Check in your code, in lexical_analyzer.py line 408, we have` while source_code[i] != "\0": `. This loops through all source_codes ("\n", "print", "(", "{", ..., "fun", ...). In there, we already check parenthesis using variable named `parantheses_count += 1`, I suggest just do the same with others braces_count, bracket_count.

Math-O5 reviewed (changes requested):
Hi @Chasmiccoder! Tnxs by the contribution. Review: Why did you create a function that will loop through all source_code again? Check in your code, in lexical_analyzer.py line 408, we have `while source_code[i] != "\0"`  This loops through all source_codes ("\n", "print", "(", "{", ..., "fun", ...). In there, we already check parenthesis using variable named parantheses_count += 1, I suggest just do the same with others braces_count, bracket_count.

Chasmiccoder commented:
Ohh.. My bad. I thought I had to perform a preliminary test on the code. @Math-O5 
I have fixed it now in the new commit. Please check it out.
Thanks!

Chasmiccoder commented:
Here are the outputs to some test cases:
Simc code:
MAIN
{
    var a = 1
    {
        var b = a+1
    }
    print(b)
END_MAIN
Error:
![image](https://user-images.githubusercontent.com/62545147/102052395-076c4d80-3e0c-11eb-8f12-f2c3b8a4a362.png)

Simc Code:
MAIN
{
    var a = 1
    {
        var b = a+1
    }
    print(b)
}
}
END_MAIN

Error:
![image](https://user-images.githubusercontent.com/62545147/102052510-2e2a8400-3e0c-11eb-84f5-7b33a024c98f.png)

When the  { } bracket is balanced, but the simple bracket is not:
MAIN
{
    var a = 1
    {
        var b = a+1
    }
    print(b))
}
END_MAIN

Error:
![image](https://user-images.githubusercontent.com/62545147/102052623-5619e780-3e0c-11eb-8476-0d0b3df5c00b.png)

frankhart2018 reviewed (changes requested):
1) Why are you looping through them? Can't the same thing be done in their respective if/elif branches? The loop adds to the complexity of the lexical_analyze function. 

2) I don't think we need three separate stacks to do the job, maintaining three stacks will not handle the error ( { ) } which should throw an error, because although the parantheses and braces are matched, they are in incorrect positions. 

3) There is a small typo in line 675, instead of openening it should be opening 😄.

Chasmiccoder commented:
1. Oh..  I thought I would save some lines of code by using a loop. I will fix that.
2. Right... did not account for that exception. I will need some time to fix it.
3. Oops.. Sorry 😓

Chasmiccoder commented:
Hi there! I have fixed the program as required:
I removed the for loop, and now the lexical analyzer accounts for balanced, but mismatched bracket sequences. 
Also, now the code uses only 1 stack.
Example:
MAIN
{
    var a = 1
    {
        var b [= a+1
    }]
    print(b)
}
END_MAIN
Output:
![image](https://user-images.githubusercontent.com/62545147/102172028-fda32280-3ebd-11eb-8cc9-aad5f35ec6ae.png)

Hope it's fine now :)

frankhart2018 reviewed (changes requested):
Why are you adding conditions parantheses/braces/brackets again? They are already there in the lexical analysis loop. This lexer will now only check whether parantheses are matched or not but will not generate any tokens for any of (, ), [, ], {, or }. Add the code to relevant conditions :).

Chasmiccoder commented:
Ohh.. Sorry @frankhart2018 .
It did not occur to me that I could append it to the already-present if statements.
I have fixed it now. Hope its fine.

frankhart2018 reviewed (changes requested):
Now the code looks perfect. Just one small change in the error message that you are throwing:-

() - Parantheses
[] - Brackets
{} - Braces

Make sure it throws appropriate error based on what's on top of the stack after lexical analysis loop.

Chasmiccoder commented:
Okay @frankhart2018 
I think I got it.
For this program:

Chasmiccoder commented:
Okay @frankhart2018 
I am not too sure how you wanted it, but I think I got it.
For this program:
MAIN
{
    var a = 1
    {
        var b [= a+1
    }]
    print(b)
}
END_MAIN
We get: 
![image](https://user-images.githubusercontent.com/62545147/102213807-4d550e80-3efd-11eb-85a1-a1ca6126650f.png)


And for:
MAIN
{{
    var a = 1
    {
        var b = a+1
    }
    print(b)
}
END_MAIN

we get:
![image](https://user-images.githubusercontent.com/62545147/102214239-f6036e00-3efd-11eb-8119-1b8b39c44afc.png)

frankhart2018 reviewed (approved):
Merging this, thanks for contributing @Chasmiccoder. Closes #402.
